pkg/apis/bindings/v1alpha1/imagebinding_lifecycle.go

```diff
@@ -26,25 +31,71 @@ func (b *ImageBinding) GetBindingStatus() duck.BindableStatus {
 	return &b.Status
 }
 
-func (b *ImageBinding) Do(ctx context.Context, ps *v1.WithPod) {
-	images := GetImages(ctx)
-	for _, p := range b.Spec.Providers {
-		for i, c := range ps.Spec.Template.Spec.Containers {
-			if c.Name == p.ContainerName {
-				img, ok := images[p.ContainerName]
-				if !ok {
-					continue
-				}
-				ps.Spec.Template.Spec.Containers[i].Image = img
-				ps.Annotations[ImageBindingAnnotationKey] = "bound"
-			}
+type imageBindingAnnotationData struct {
+	Container string `json:"container"`
+	Provider  string `json:"provider"`
+}
+
+func (b *ImageBinding) annotationKey() string {
+	return ImageBindingAnnotationKeyPrefix + b.Name
+}
+
+func (b *ImageBinding) annotationVal() string {
+	gvr, _ := meta.UnsafeGuessKindToResource(b.Spec.Provider.GroupVersionKind())
+	ba := &imageBindingAnnotationData{
+		Container: b.Spec.ContainerName,
+		Provider:  gvr.GroupResource().String() + "/" + b.Spec.Provider.Name,
+	}
+	data, _ := jsoniter.MarshalToString(ba)
+	return data
+}
+
+func (b *ImageBinding) Do(ctx context.Context, ps *duckv1.WithPod) {
+	latestImage := GetLatestImage(ctx)
+	if latestImage == "" {
+		return
+	}
+	b.Undo(ctx, ps)
+	b.doContainers(ps.Spec.Template.Spec.Containers, ps, latestImage)
+	b.doContainers(ps.Spec.Template.Spec.InitContainers, ps, latestImage)
+}
+
+func (b *ImageBinding) doContainers(ctrs []corev1.Container, ps *duckv1.WithPod, latestImage string) {
+	for i, c := range ctrs {
+		if c.Name != b.Spec.ContainerName {
+			continue
+		}
+		if belongsToAnotherBinding(c.Name, ps) {
+			return
+		}
+		ctrs[i].Image = latestImage
+		if ps.Annotations == nil {
+			ps.Annotations = map[string]string{}
+		}
+		ps.Annotations[b.annotationKey()] = b.annotationVal()
+	}
+}
+
+func belongsToAnotherBinding(container string, ps *duckv1.WithPod) bool {
+	for k, v := range ps.Annotations {
```

Comment:
no need to iterate over all annotations if the container name and binding name were flipped